### ISSUE TITLE
Feature Request: "Play" Button to Auto-Blink All Frames?

### ISSUE THREAD

**Issue by pllim** [ejeschke/ginga#120]
Just an idea. On DS9, there is a feature where you press a button called "Blink" (under "Frames") and you can automatically blink through all visible frames. You also have the option to set the blink interval (0.125, 0.25, 0.5, 1, 2, 4, or 8 seconds). 

Somewhat related (but maybe not as important): There is also a feature on DS9 called "Create Movie..." (http://ds9.si.edu/doc/ref/command.html#movie) that can capture the blinking in a form that is shareable as a `.mpg` file.

**Comment by ejeschke** [comment]
I think we should add such a feature, because I have received all kinds of verbal queries about it.  I'm not sure how straightforward it is to implement subsecond timing accuracy switching between tabs, but there is some support already for this kind of switching.  If you press 'n' in the window (of the reference viewer) it should switch to the next tab and it will cycle around when you get to the end.  @pllim , can you try this and let me know if it seems like a reasonable facimile of the "blink" feature?  If so we could implement a little timer to automate it on a chosen time.

**Comment by pllim** [comment]
Nothing happens when I press 'n'. I don't see any binding for 'n' in your `examples/bindings/bindings.cfg.standard` either, which I assume is the default binding?

**Comment by ejeschke** [comment]
'n' is not a widget binding.  There are a handful of keystrokes that are only handled by the reference viewer and so are not in the bindings.cfg (getting them in there is on the todo list).  These can be found in  [Control.py](https://github.com/ejeschke/ginga/blob/master/ginga/Control.py) in the `keypress` method.

I might not have been clear about the operation.  You need to have two or more channels (and therefore tabs) open.  Pressing 'n' will toggle through the tabs.  Click in one of the channel windows just to make sure that the widget has the focus and press 'n'.  Does this not cycle the tabs?

**Comment by pllim** [comment]
I think we're talking about two different things here. What I want is auto-play for all the images in the same channel. In my case, 'n' won't help.

**Comment by ejeschke** [comment]
Hmm, Ok, it seems this is a matter of workflow, maybe.  In any case, it also should not be hard to add a feature to switch between several images in a channel's datasrc.  Since the datasrc can hold an arbitrary number of images, should the feature just flip between the current one and the previous one, or all of them in the datasrc?

**Comment by ejeschke** [comment]
Could you try using the left arrow/right arrow (keyboard) keys to go back and forth between two images?  Is this basically what you have in mind, but controlled by a timer?

**Comment by pllim** [comment]
> should the feature just flip between the current one and the previous one, or all of them in the datasrc?

All of them. DS9 also has "show/hide", so it skips the hidden ones, but that can be added in the future if needed. Come to think of it, there also should be a "stop" button to go with "play".

> Is this basically what you have in mind, but controlled by a timer?

Yes.

**Comment by ejeschke** [comment]
@pllim, could you take a look at commit 
commit 69362a2b7bfadf2385697d8194e0f131448bc93e?

You will find a new "Blink" plugin in the Operation menu.  Let me know how it works for you.

**Comment by pllim** [comment]
@ejeschke , I cannot find the "Blink" plugin and I also don't see any `Blink.py` in your commit.

**Comment by ejeschke** [comment]
Oops, forgot to commit the new plugin.  Please check commit aadfab9856054360fca2bc54bc806421cbc5dc7e.

**Comment by pllim** [comment]
As mentioned in #119, only the images still in memory are included in the blinking, not all the images listed under `Thumbs` or `Contents`.

Regarding the plugin GUI:
*  Interval needs a tool tip and a text describing what time unit the value is in. I'm guessing it is in seconds?
* The buttons need to be the same size, which probably means they need to go in a separate container than the one containing the interval stuff.

**Comment by ejeschke** [comment]
I think if the blink is supposed to include images that are not in memory this would be difficult to time accurately, as it would be swapping them from disk repeatedly.  I suppose ds9 only blinks images that are loaded/in memory?

Any reason not to increase the size of the channel's `Datasrc` to be a size necessary to hold all of the images?  Maybe this length needs to be able to be adjusted dynamically from Preferences?

**Comment by pllim** [comment]
>  I suppose ds9 only blinks images that are loaded/in memory?

Probably, since DS9 has a max limit (16?) on the number of displayed frames.

> Any reason not to increase the size ...?

No, but consider this situation from the viewpoint of a typical user who isn't aware of all the configurations and behind the scene stuff: User does not have any custom configuration (so the default `numImages` is used). User opens a bunch of images that exceeded `numImages` but user does not even know that `numImages` exists. User blinks through the images, expecting the blinking to go through all the images, but is very confused when some appear missing. It is fine if you only want `Blink` to read from buffer, but the GUI needs to be clear about this limitation.

**Comment by ejeschke** [comment]
This raises an interesting question which is sort of related to the problem of recreating the mosaics when they have been ejected.  The reference viewer was created initially in an environment where data is cycling through the viewer all night long. The Thumbs and Contents hide the bounded nature of the channels `Datasrc`s, allowing the old data to be loaded with a slightly longer delay.

Perhaps we need an option to make `Datasrc`s unbounded.  That puts the onus on the user to manage their own memory by removing images that are no longer necessary.  What do you think?

**Comment by pllim** [comment]
Hmm. I do like the bounded `Datasrc` feature. I never liked the 16-frame limit on DS9, so I can't say I fully support the unbounded `Datasrc` idea. This will be a significant change, so you might want to ask feedback from more people.

**Comment by ejeschke** [comment]
@pllim, if you check out commit dc1eccb7ea48940791e1aaa783d2b6466ff96a75, I have improved the button layout and added a tool tip as requested.

It's maybe not the best solution, but you can now change the limit on the `Datasrc` in Preferences. If you set it to 0 you will get an unlimited buffer.

**Comment by ejeschke** [comment]
@pllim, unrelated to this issue, but try out the revamped "Cuts" plugin in this commit.  Let me know what you think.

**Comment by pllim** [comment]
Thanks for the updates!

* I set `numImages` to 0 in `Preferences` and click "Save".
  * When I check `channel_Image.cfg` and `general.cfg` in my `$HOME/.ginga` directory, I still see `numImages=4`
  * Changing the value in `Preferences` without restarting Ginga did not make `Blink` load the rest of the images listed under `Contents`

* `Blink` GUI looks good

* I played with `Cuts` but drawing line, path, and free path repeatedly, and press "Delete" once in a while.
  * In the left-most drop-down box, I see "None" and "CutX", but there is always only one "CutX", not multiple even after I repeatedly switched between line, path, and free path
  * "X" in "CutX" does not reset after I press "Delete", but it keeps incrementing, which is a bit unexpected but no big deal
  * I get the following `RuntimeWarning` but no error in Ginga log file
```
.../ginga-2.4.20150626080455-py2.7.egg/ginga/canvas/CanvasObject.py:539: RuntimeWarning: invalid value encountered in double_scalars
  Cx = numpy.sum(cx * (a - b)) / (6. * A)
```

**Comment by ejeschke** [comment]
I just tested saving the Preferences and in my case the numImages *did* change in the file.  But I pressed Enter after entering the value in the field.  Could you see whether that works for you?  I'll try to fix it to not require the Enter.

**Comment by ejeschke** [comment]
Cuts GUI or instructions probably need to be improved.  In the current operation style, you select None if you want to draw a new cut (or to select a cut for Delete).  To keep redrawing the same cut set it to something other than None.  The type drop down just chooses what kind of cut you are going to draw.

**Comment by ejeschke** [comment]
Hmm, let me rephrase that: 
 In the current operation style, you select None if you want to draw a new cut. Otherwise, if a particular named cut is selected in the drop down box then it will be replaced by a new drawn cut.  To delete a cut select its name and press Delete. The type drop down just chooses what kind of cut you are going to draw.

**Comment by pllim** [comment]
> But I pressed Enter after entering the value in the field. Could you see whether that works for you?

Yes, pressing Enter saved `numImages=0` in `channel_Image.cfg`. *But* it does not make `Blink` to cycle through all the images listed in `Contents`, which implies that `Datasrc` is not picking up this change without having to quit and then restart Ginga.

**Comment by pllim** [comment]
> Cuts GUI or instructions probably need to be improved.

The information you mentioned above would be useful in "Instructions".

Other minor GUI comments:

* When a cut is Drawn, is it possible to set the left drop-down menu back to "None"? Currently, the menu remains on the latest cut, which defaults to the behavior of overwriting the cut when I draw another line. If I forget to re-select "None" before drawing again, I lose my latest cut. Does that make sense?
* Delete buttons can be disabled when there is nothing to delete. Right now, they just do nothing when you press them anyway, so this is more as a cosmetic improvement.
* Plot should clear completely when the last cut is deleted, or when "Delete All" is pressed. Right now, the last cut is displayed if they are deleted one by one, or the whole plot remains if they are all deleted at once. The remaining plot is not meaningful because it is no longer associated with any cut lines. Again, a cosmetic improvement.
* I notice that the cut color changes when I delete another cut. Personally, I am fine with it, but I don't know if other users will find it confusing. For example, Cut1 is green, Cut2 is red; Then, I delete Cut1, and now Cut2 is green.

**Comment by pllim** [comment]
I set `numImages=0` in my `general.cfg` file and `Blink` appears to cycle through all the images now. But how do I make it such that it cycles in alphabetical order sorted by image name? Is that possible? Or at least in the same order as appear on `Contents`? Such sorting is also desirable for when the green arrows are pressed. IMHO, green arrows and `Blink` should have the same behavior, because theoretically `Blink` is just the arrows automated.

Would be visually helpful if the correct row on `Contents` is highlighted when `Blink` is running, or when one of the green arrows is pressed. This is because when a row is selected (and highlighted) in `Contents`, the main display changes to the selected row. So in reverse, when the display changes (e.g., by using `Blink` or green arrow), the correct row should in turn be highlighted in `Contents`, no?

**Comment by pllim** [comment]
Side note: Setting `numImages` in `Preferences` works correctly when I press Enter, as you suggested. That's fine but should be documented somehow on the GUI (maybe tooltip?), e.g., "press Enter to set new value" or something like that. Pressing Enter is not intuitive to me because I already see the number changed on GUI without pressing Enter.

**Comment by ejeschke** [comment]
Actually, there are a number of text entry boxes in `Preferences` in which one must press enter to "commit" the value.  I understand the issue of not understanding when the value will be picked up or not.  In some places (e.g. cut levels controls in `Synopsis` tab) you can see one approach I take to this uncertainly which is to show the actual values to the left of the controls--these will change only when Enter is pressed or the `Cut Levels` button is pressed.  Perhaps it may be best to simply have a button next to every such text box throughout the program?  Not sure what the correct solution is yet.

**Comment by ejeschke** [comment]
Try commit 0f9d430fca770c6b87f13aeb93ebfc05cdacef2d when you get a chance.  It adds a widget type that has a "Set" button next to the entry field.  It's used in a couple of places in `Preferences`.  Let me know what you think.

**Comment by ejeschke** [comment]
@pllim, I'm wondering if a rethink is in order for the `Blink` plugin to allow better control over *which* images are selected for the blink and what order they are shown.  I'm thinking there could be a reorderable list widget that would show all the images in memory in the channel.  You could drag to reorder and check a checkbox next to the ones you want to blink.  We could even provide a control for dynamically changing the channel buffer length (`numImages`).  What do you think?

**Comment by pllim** [comment]
The list for reordering would be nice. Checkboxes would be a bonus, and would be analogous to show/hide in DS9. The trick here is to have the "green arrows" (for manual blinking) pick these up as well. Otherwise, it would be confusing to have `Blink` plugin behave one way, but the built-in green arrow another way.

Personally, I wouldn't find dynamic change of `numImages` useful because I always set `numImages=0` (and I wouldn't want Ginga to change that).

**Comment by pllim** [comment]
@ejeschke , the "Set" button makes it more obvious, thanks! Some kind of text feedback would be nice too (if you feel like implementing it), i.e., a "status bar" right above the "Close" and "Save" buttons that would say something like "Num Images changed from 10 to 0" or "XXX change from old value to new value" if another preference is changed.

**Comment by pllim** [comment]
This is resolved by #226. There is a minor bug that should be fixed by #238.
